include/shuffle/BijectiveFunctionScanShuffle.h

```diff
@@ -23,34 +29,36 @@ class BijectiveFunctionScanShuffle : public Shuffle<ContainerType, RandomGenerat
         uint64_t capacity = mapping_function.getMappingRange();
 
         thrust::counting_iterator<uint64_t> indexes( 0 );
-        thrust::device_vector<uint64_t> mappings( capacity );
-        auto mapping_it = mappings.begin();
-        thrust::transform( indexes, indexes + capacity, mapping_it,
-                           [mapping_function] __host__ __device__( uint64_t val ) -> uint64_t {
-                               // Call mapping functions
-                               return mapping_function( val );
-                           } );
-
-        // Transform iterator for low memory footprint - worse performance
-        // auto mapping_it =
-        //     thrust::make_transform_iterator( indexes, [mapping_function] __host__ __device__( uint64_t val ) -> uint64_t {
-        //         // Call mapping functions
-        //         return mapping_function( val );
-        //     } );
-
-        auto is_valid_it =
-            thrust::make_transform_iterator( mapping_it, [num] __host__ __device__( uint64_t val ) -> bool {
-                // Check if mapping is valid
-                return val < num;
-            } );
-
-        thrust::device_vector<uint64_t> scan_buf( capacity );
-        thrust::exclusive_scan( is_valid_it, is_valid_it + capacity, scan_buf.begin() );
-        auto out_permutation_iterator =
-            thrust::make_permutation_iterator( out_container.begin(), scan_buf.begin() );
-
-        thrust::gather_if( mapping_it, mapping_it + capacity, is_valid_it, in_container.begin(),
-                           out_permutation_iterator );
+        size_t m = num;
+        auto input_iter = in_container.begin();
+        auto mapping_it = thrust::make_transform_iterator( indexes, [=] __device__( uint64_t val ) -> ValueT {
+            auto gather_key = mapping_function( val );
+            if( gather_key < m )
+            {
+                return input_iter[gather_key];
+            }
+            return ValueT();
+        } );
+        auto flag_it = thrust::make_transform_iterator( indexes, [=] __device__( uint64_t val ) {
+            auto gather_key = mapping_function( val );
```

Comment:
Why are we running the function twice?

Reply:
Maybe we could do an initial transform into a second temp buffer to avoid this?

Reply:
In principle it can work with 1 function evaluation, but I had problems making output iterators work with cub correctly. I can try again later if there is a visible perf difference. 

It's important to me that we avoid the temporary buffer for better memory usage.

If you do it in 2 passes you also pay 2x for memory loads. In cuda memory loads tend to dominate arithmetic for most things.

Reply:
It might be worth adding a static counter to see how many times the function is run anyway. We might also be able to return a pair from the first and use a transform iterator to get a specific element from the pair (though that would have the same result if the transform iterator doesn't cache)